src/main/kotlin/com/zepben/evolve/services/network/EquivalentNetworkUtils.kt

```diff
@@ -0,0 +1,465 @@
+     * @param otherContainerClass The edge [EquipmentContainer] class that doesn't have its equipment in the [NetworkService] (it holds the edge [ConnectivityNode]).
+     * @param network The [NetworkService] that we need to add the [EquivalentBranch] and [ConductingEquipment] to.
+     * @param createEquivalentBranches factory function to create the [EquivalentBranch]es that will be attached to the edge [ConnectivityNode] to represent the 'equivalent network'.
+     * @param createEquivalentEquipment factory function to create the [ConductingEquipment]s that will be attached to the [EquivalentBranch]es of the 'equivalent network'.
+     * @param maxNumber The maximum number of 'equivalent networks' you want to add.
+     *
+     * @return a set of [EquivalentNetworkConnection]s that hold information and references to the edge [ConnectivityNode], edge [ConductingEquipment], and added [EquivalentBranch]es and their [ConductingEquipment].
+     */
+    fun addToEdgeBetweenContainers(
+        containerClass: KClass<out EquipmentContainer>,
+        otherContainerClass: KClass<out EquipmentContainer>,
+        network: NetworkService,
+        createEquivalentBranches: (EquivalentBranchDetails) -> Sequence<Pair<PhaseCode, EquivalentBranch>>,
+        createEquivalentEquipment: (EquivalentEquipmentDetails) -> Sequence<Pair<PhaseCode, ConductingEquipment>>,
+        maxNumber: Int? = null
+    ): Set<EquivalentNetworkConnection> {
+        val edgeCnn = getEdgeNodes(network, setOf(containerClass, otherContainerClass))
+        val edgeCnToProcess = (maxNumber?.let { edgeCnn.take(it) } ?: edgeCnn)
+        return edgeCnToProcess.map { cn -> addEquivalentNetwork(network, cn, createEquivalentBranches, createEquivalentEquipment) }.toSet()
+    }
+
+    /**
+     * Convenience function to get the first [PhaseCode] from a [ConnectivityNode]s [Terminal]s.
+     */
+    fun getPhaseCode(edgeNode: ConnectivityNode): PhaseCode = edgeNode.terminals.map { it.phases }.first()
+
+    /**
+     * Convenience function to get the [BaseVoltage] from an edge [ConductingEquipment].
+     */
+    fun getEquivalentBaseVoltage(ce: ConductingEquipment, edgeNode: ConnectivityNode): BaseVoltage? =
+        when (ce) {
+            is PowerTransformer -> ce.ends.first { it.terminal?.connectivityNode == edgeNode }.baseVoltage
+            else -> ce.baseVoltage
+        }
+
+    private fun addEquivalentNetwork(
+        network: NetworkService,
+        edgeNode: ConnectivityNode,
+        createEquivalentBranches: (EquivalentBranchDetails) -> Sequence<Pair<PhaseCode, EquivalentBranch>>,
+        createEquivalentEquipment: (EquivalentEquipmentDetails) -> Sequence<Pair<PhaseCode, ConductingEquipment>>
+    ): EquivalentNetworkConnection {
+        val edgeEquipment = edgeNode.terminals.first().conductingEquipment
+            ?: throw IllegalStateException("ConnectivityNode for EquivalentBranch must have ConductingEquipment")
```

Comment:
I know this exception shouldn't occur, but can you add more details here. You probably want to include the `edgeNode.mRID` and the terminals mRID in the message

Reply:
Yep, it should never happen but added a comment and improved the message in the exception.